p2p/peersHolder/peersHolder.go

```diff
@@ -0,0 +1,199 @@
+package peersHolder
+
+import (
+	"strings"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+)
+
+type peerInfo struct {
+	pid     core.PeerID
+	shardID uint32
+}
+
+type peerIDData struct {
+	connectionAddressSlice string
+	shardID                uint32
+	index                  int
+}
+
+type peersHolder struct {
+	preferredConnAddrSlices    []string
+	connAddrSliceToPeerInfo    map[string]*peerInfo
+	tempPeerIDsWaitingForShard map[core.PeerID]string
+	peerIDsPerShard            map[uint32][]core.PeerID
+	peerIDs                    map[core.PeerID]*peerIDData
+	sync.RWMutex
+}
+
+// NewPeersHolder returns a new instance of peersHolder
+func NewPeersHolder(preferredConnectionAddressSlices [][]byte) *peersHolder {
+	preferredConnections := make([]string, 0)
+	connAddrSliceToPeerIDs := make(map[string]*peerInfo)
+
+	for _, connAddrSlice := range preferredConnectionAddressSlices {
+		preferredConnections = append(preferredConnections, string(connAddrSlice))
+		connAddrSliceToPeerIDs[string(connAddrSlice)] = nil
+	}
+
+	return &peersHolder{
+		preferredConnAddrSlices:    preferredConnections,
+		connAddrSliceToPeerInfo:    connAddrSliceToPeerIDs,
+		tempPeerIDsWaitingForShard: make(map[core.PeerID]string),
+		peerIDsPerShard:            make(map[uint32][]core.PeerID),
+		peerIDs:                    make(map[core.PeerID]*peerIDData),
+	}
+}
+
+// PutConnectionAddress will perform the insert or the upgrade operation if the provided peerID is inside the preferred peers list
+func (ph *peersHolder) PutConnectionAddress(peerID core.PeerID, connectionAddress []byte) {
+	ph.Lock()
+	defer ph.Unlock()
+
+	knownSlice := ph.getKnownSlice(string(connectionAddress))
+	if len(knownSlice) == 0 {
+		return
+	}
+
+	pInfo := ph.connAddrSliceToPeerInfo[knownSlice]
+	if pInfo == nil {
+		ph.tempPeerIDsWaitingForShard[peerID] = knownSlice
+		ph.connAddrSliceToPeerInfo[knownSlice] = &peerInfo{
+			pid:     peerID,
+			shardID: 0, // this will be overwritten once shard is available
+		}
+
+		return
+	}
+
+	isOldData := peerID == pInfo.pid
```

Comment:
how is this gonna perform if you have defined a preferred connection string as an IP but that IP has 3 peer IDs running?

Reply:
updated the map to support multiple pids for the same connection addr